TACC/tup-ui
# task/TUP-194: Integrate dashboard login with Django user model

## Overview:
Integrates TUP-UI login with the TUP-CMS user model. The flow works like this: 
1. The user tries to visit a protected route (anything other than `/dashboard/login` for now).
2. They get bounced and redirected to `/dashboard/login` if their `x-tup-token` cookie is missing/invalid.
3. Upon authenticating, they are redirected to the dashboard using `window.location.replace()`, causing the Django authentication background to login their user (or create it if it doesn't exist).

## Related:

- [TUP-194](https://jira.tacc.utexas.edu/browse/TUP-194)

## Changes:
- Refactor `LoginComponent` to use the `window.location` API to redirect instead of the history API. This is necessary for the Django backend to recognize the login event.
- Add a custom authentication backend that checks the user's `x-tup-token` cookie and validates it against the TUP-services API.
- Add a new `RestrictedView` that attempts to authenticate the user and redirects to `/dashboard/login` otherwise. 

## Todo:
- Make redirects work for custom apps other than the dashboard (nothing will break now but it isn't smart about redirecting users to the specific page they were trying to access).

## Testing:

1. Navigate to `localhost:8000/dashboard/` and confirm that you are redirected to a login page.
2. Log in using your TACC credentials and confirm that you are redirected to your profile page.
3. Log out and log back in to check that the auth flow works as expected.
